### PR TITLE
Identity refactor pipeline

### PR DESCRIPTION
**Pull request by billwert**
Refactor the implementation of `AzurePipelinesCredential` to use more of the shared infrastructure in `Identity(Sync)Client` without creating a second `HttpPipeline`.

Also preps for release.

### CODE REVIEW

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/AzurePipelinesCredential.java [hunk 1]
```diff
@@ -63,41 +65,65 @@ public class AzurePipelinesCredential implements TokenCredential {
...
-                    }
-                    try (JsonReader reader = JsonProviders.createReader(responseBody)) {
-                        return OidcTokenResponse.fromJson(reader).getOidcToken();
+
+        IdentityClientBuilder builder = new IdentityClientBuilder()
+            .tenantId(tenantId)
+            .clientId(clientId)
+            .identityClientOptions(identityClientOptions)
+            .clientAssertionSupplierWithHttpPipeline((httpPipeline) -> {
+                try {
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/AzurePipelinesCredential.java:74]: nit: This function can be refactored to a method in IdentityUtil to improve readability of landing public class.

Comment (billwert) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/AzurePipelinesCredential.java:74]: I think you mean the callback, yeah? I don't think it belongs somewhere else; this type defining it seems very reasonable to me. I could be convinced to put it in a method on this type and pass that method reference in if you felt strongly about this much code in the ctor.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java [hunk 2]
```diff
@@ -129,7 +129,7 @@ public abstract class IdentityClientBase {
     private static final ClientOptions DEFAULT_CLIENT_OPTIONS = new ClientOptions();
 
 
-    private static final Map<String, String> PROPERTIES = CoreUtils.getProperties(AZURE_IDENTITY_PROPERTIES);
+    private final Map<String, String> properties = CoreUtils.getProperties(AZURE_IDENTITY_PROPERTIES);
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java:132]: why make this non static ? 
doesn't need to be instantiated per instance.

Comment (billwert) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java:132]: It previously was non-static before I made the change that exposed `setupPipeline`. See [here](https://github.com/Azure/azure-sdk-for-java/pull/40246/files#diff-b5937fc0e636a1c6954f1b77ed7149066d1f7048dd85d8db025609d2615c514cR130). I just rolled back all the things involved in making `setupPipeline` a static usable outside the class.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java [hunk 3]
```diff
@@ -886,16 +902,13 @@ void initializeHttpPipelineAdapter() {
...
         } else {
             // If http client is set on the credential, then it should override the proxy options if any configured.
             HttpClient httpClient = options.getHttpClient();
-            if (httpClient != null) {
-                httpPipelineAdapter = new HttpPipelineAdapter(setupPipeline(httpClient, this.options), options);
-            } else if (options.getProxyOptions() == null) {
-                //Http Client is null, proxy options are not set, use the default client and build the pipeline.
-                httpPipelineAdapter = new HttpPipelineAdapter(setupPipeline(HttpClient.createDefault(), options), options);
-            }
+            this.httpPipeline = setupPipeline(httpClient == null ? HttpClient.createDefault() : httpClient);
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java:910]: this breaks the proxy flow I think (backward compat), the Pipeline adapter in current flow is only configured if either httpClient is not null or proxy options are not configured.

With this change, the pipeline adapter will get configured if httpClient is null and proxy options are configured.

Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java:910]: This call will break with this change. With proxy options (backward compat), this flow is currently invoked.

https://github.com/Azure/azure-sdk-for-java/pull/40722/files#diff-b5937fc0e636a1c6954f1b77ed7149066d1f7048dd85d8db025609d2615c514cR277

To retain this, I'd say, we move this logic up, and before setting client Assertion credentials, check if pipeline is null then create the pipeline there for the assertion to consume.

Comment (billwert) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java:910]: good catch, fixing!

**Review by g2vinay — APPROVED** [review]

**Pull request merged**